repository/packer/packwriter_test.go

```diff
@@ -0,0 +1,879 @@
+		for i := 0; i < 5; i++ {
+			mac := objects.RandomMAC()
+			data := []byte("test data " + string(rune(i+'0')))
+			err := pw.WriteBlob(resources.RT_CONFIG, versioning.FromString("1.0.0"), mac, data, uint32(i))
+			if err != nil {
+				return err
+			}
+		}
+		err := pw.Finalize()
+		require.NoError(t, err)
+		return nil
+	}
+	pw := NewPackWriter(putter, encoder, hasher, packingCache)
+	require.NotNil(t, pw)
+}
+
+func TestPackWriter_InternalStateAccess(t *testing.T) {
+	tmpDir := t.TempDir()
+	cacheManager := caching.NewManager(tmpDir)
+	defer cacheManager.Close()
+	packingCache, err := cacheManager.Packing()
+	require.NoError(t, err)
+	defer packingCache.Close()
+	encoder := func(r io.Reader) (io.Reader, error) { return r, nil }
+	hasher := func() hash.Hash { return sha256.New() }
+	putter := func(pw *PackWriter) error {
+		// Test access to internal fields
+		require.NotNil(t, pw.hasher)
+		require.NotNil(t, pw.writer)
+		require.Equal(t, uint64(0), pw.currentOffset)
+		require.Equal(t, uint32(0), pw.Footer.Count)
+
+		mac := objects.RandomMAC()
+		data := []byte("test data")
+		err := pw.WriteBlob(resources.RT_CONFIG, versioning.FromString("1.0.0"), mac, data, 0)
+		if err != nil {
+			return err
+		}
+
+		require.Equal(t, uint32(1), pw.Footer.Count)
+		require.Greater(t, pw.currentOffset, uint64(0))
+		return nil
+	}
+	pw := NewPackWriter(putter, encoder, hasher, packingCache)
+	require.NotNil(t, pw)
+}
+
+func TestPackWriter_ErrorHandlingInternal(t *testing.T) {
+	tmpDir := t.TempDir()
+	cacheManager := caching.NewManager(tmpDir)
+	defer cacheManager.Close()
+	packingCache, err := cacheManager.Packing()
+	require.NoError(t, err)
+	defer packingCache.Close()
+	encoder := func(r io.Reader) (io.Reader, error) { return r, nil }
+	hasher := func() hash.Hash { return sha256.New() }
+	putter := func(pw *PackWriter) error {
+		// Test error handling in writeAndSum
+		// Create a writer that will fail
+		failingWriter := &failingWriter{}
+		err := pw.writeAndSum(failingWriter, uint32(0x12345678))
+		require.Error(t, err)
+		return nil
+	}
+	pw := NewPackWriter(putter, encoder, hasher, packingCache)
+	require.NotNil(t, pw)
+}
+
+type failingWriter struct{}
+
+func (fw *failingWriter) Write(p []byte) (n int, err error) {
+	return 0, io.ErrShortWrite
+}
+
+func TestPackWriter_ConcurrentInternalAccess(t *testing.T) {
+	tmpDir := t.TempDir()
+	cacheManager := caching.NewManager(tmpDir)
+	defer cacheManager.Close()
+	packingCache, err := cacheManager.Packing()
+	require.NoError(t, err)
+	defer packingCache.Close()
+	encoder := func(r io.Reader) (io.Reader, error) { return r, nil }
+	hasher := func() hash.Hash { return sha256.New() }
+	putter := func(pw *PackWriter) error {
+		done := make(chan error, 3)
+		for i := 0; i < 3; i++ {
+			go func(index int) {
+				var buf bytes.Buffer
+				err := pw.writeAndSum(&buf, uint32(index))
+				done <- err
+			}(i)
+		}
+
+		for i := 0; i < 3; i++ {
+			err := <-done
+			if err != nil {
+				return err
+			}
+		}
```

Comment:
one minor nit: if `pw.writeAndSum` fails (assuming it can), then in the error we might stop at the first iteration and leave the other two goroutines waiting indefinitely on `done <- err`.

I don't think it's a big deal for tests, but can't we use an errgroup here?  it should take care of that automatically, or just set the err but continue reading.

Reply:
thanks ! fixed in https://github.com/PlakarKorp/kloset/pull/51/commits/eb177cd565ed9402d5f215182738eb99a6ea266b